File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
# Time Complexity: O(2*log(n))
# Space Complexity: O(1)
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def binary_search(nums, target, is_low):
            low, high = 0, len(nums)-1
            result_idx = -1
            while low <= high:
                mid = (low + high)//2
                if nums[mid] == target:
                    result_idx = mid
                    if is_low:
                        high = mid - 1
                    else:
                        low = mid + 1
                elif nums[mid] < target:
                    low = mid + 1
                else:
                    high = mid - 1
            return result_idx
        
        low = binary_search(nums, target, True)
        high = binary_search(nums, target, False)

        return [low, high]

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.rs
// Time Complexity: O(log(N))
// Space Complexity: O(1)
impl Solution {
    
    pub fn binary_search(nums: &Vec<i32>, key: i32, is_low: bool) -> i32 {
        let mut left = 0;
        let mut right = nums.len() as i32 - 1;
        let mut result = -1; // Store the result index

        while left <= right {
            let mid = left + (right - left) / 2; // Avoid potential overflow
            if nums[mid as usize] == key {
                result = mid; // Update result when key is found
                if is_low {
                    right = mid - 1; // Narrow search to the left for lower bound
                } else {
                    left = mid + 1; // Narrow search to the right for upper bound
                }
            } else if nums[mid as usize] > key {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        result
    }

    pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {
        if nums.len() == 0 {
            return Vec::<i32>::from([-1, -1]);
        }
        let start_pos = Self::binary_search(&nums, target, true);
        let end_pos = Self::binary_search(&nums, target, false);
        Vec::<i32>::from([start_pos, end_pos])
    }
}